cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,710 @@
+        ? ((*segment_offsets).size() > (detail::num_sparse_segments_per_vertex_partition + 1))
+        : false;
+    if (use_dcs) {
+      auto major_hypersparse_first =
+        matrix_partition.get_major_first() +
+        (*segment_offsets)[detail::num_sparse_segments_per_vertex_partition];
+      hypersparse_begin.push_back(major_hypersparse_first);
+    } else {
+      hypersparse_begin.push_back(graph_view.get_local_adj_matrix_partition_row_last(i));
+    }
+
+    counter += matrix_partition.get_major_size();
+  }
+
+  // Allocate device memory for transfer
+  rmm::device_uvector<partition_t> matrix_partitions(
+    graph_view.get_number_of_local_adj_matrix_partitions(), handle.get_stream());
+
+  rmm::device_uvector<vertex_t> r_begin(id_begin.size(), handle.get_stream());
+  rmm::device_uvector<vertex_t> r_end(id_end.size(), handle.get_stream());
+  rmm::device_uvector<vertex_t> vc_offsets(vertex_count_offsets.size(), handle.get_stream());
+  rmm::device_uvector<vertex_t> hs_begin(hypersparse_begin.size(), handle.get_stream());
+
+  // Transfer data
+  raft::update_device(
+    matrix_partitions.data(), partitions.data(), partitions.size(), handle.get_stream());
+  raft::update_device(r_begin.data(), id_begin.data(), id_begin.size(), handle.get_stream());
+  raft::update_device(r_end.data(), id_end.data(), id_end.size(), handle.get_stream());
+  raft::update_device(vc_offsets.data(),
+                      vertex_count_offsets.data(),
+                      vertex_count_offsets.size(),
+                      handle.get_stream());
+  raft::update_device(
+    hs_begin.data(), hypersparse_begin.data(), hypersparse_begin.size(), handle.get_stream());
+
+  return std::make_tuple(std::move(matrix_partitions),
+                         std::move(r_begin),
+                         std::move(r_end),
+                         std::move(hs_begin),
+                         std::move(vc_offsets));
+}
+
+}  // namespace detail
+
+/**
+ * @brief Return global out degrees of active sources
+ *
+ * Get partition information of all graph partitions on the gpu and select
+ * global degrees of all active sources
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param active_sources Device vector containing all the vertex id that are processed by
+ * gpus in the column communicator
+ * @param global_out_degrees Global out degrees for every source represented by current gpu
+ * @return Global out degrees of all sources in active_sources
+ */
+template <typename GraphViewType>
+rmm::device_uvector<typename GraphViewType::edge_type> get_active_source_global_degrees(
+  raft::handle_t const& handle,
+  GraphViewType const& graph_view,
+  rmm::device_uvector<typename GraphViewType::vertex_type>& active_sources,
```

Comment:
Can't this be `const`?

And if this is read-only, now we should consider using `raft::device_span` (and actually, if we don't need to resize, we can use device_span).

Reply:
I was in the process of changing it when I found out that you have to explicitly create a span out of the device_uvector. Perhaps we should add a constructor that instantiates from supported containers like cudf.

Reply:
I see, not a high priority thing at all, if this is a tricky thing, then, we can revisit this later in a separate PR (maybe create a PR just updating cugraph public C++ API functions to use span).